tests/installation/npmTest.ts

```diff
@@ -0,0 +1,387 @@
+        },
+      },
+    };
+
+    const object = path.join(this._objectsDir, `${shasum}.tgz`);
+    await fs.promises.copyFile(tarPath, object);
+    this._packageMeta.set(pkg, [metadata, object]);
+  }
+
+  private _logAccess(info: {status: 'PROXIED' | 'LOCAL', pkg: string, type?: 'tar' | 'metadata'}) {
+    this._log.push(info);
+  }
+}
+
+export type SpawnResult = {
+  stdout: string,
+  stderr: string,
+  code: number,
+  error?: any,
+};
+
+export class ExecOutput {
+  public readonly raw: SpawnResult;
+
+  constructor(result: SpawnResult) {
+    this.raw = result;
+  }
+
+  combined() {
+    return `${this.raw.stdout}\n${this.raw.stderr}\n`;
+  }
+}
+
+export type ExecOptions = { cwd?: string, env?: Record<string, string>, message?: string, expectToExitWithError?: boolean };
+export type ArgsOrOptions = [] | [...string[]] | [...string[], ExecOptions] | [ExecOptions];
+export const test = _test.extend<{
+    _autoCopyScripts: void,
+    tmpWorkspace: string,
+    nodeMajorVersion: number,
+    installedBrowsers: () => Promise<string[]>;
+    writeFiles: (nameToContents: Record<string, string>) => Promise<void>,
+    exec: (cmd: string, ...argsAndOrOptions: ArgsOrOptions) => Promise<string>
+    tsc: (...argsAndOrOptions: ArgsOrOptions) => Promise<string>,
+    registry: Registry,
+        }>({
+          _autoCopyScripts: [async ({ tmpWorkspace }, use) => {
+            const dstDir = path.join(tmpWorkspace);
+            const sourceDir = path.join(__dirname, 'fixture-scripts');
+            const contents = await fs.promises.readdir(sourceDir);
+            await Promise.all(contents.map(f => fs.promises.copyFile(path.join(sourceDir, f), path.join(dstDir, f))));
+            await use();
+          }, {
+            auto: true,
+          }],
+          nodeMajorVersion: async ({}, use) => {
+            await use(+process.versions.node.split('.')[0]);
+          },
+          writeFiles: async ({ tmpWorkspace }, use) => {
+            await use(async (nameToContents: Record<string, string>) => {
+              for (const [name, contents] of Object.entries(nameToContents))
+                await fs.promises.writeFile(path.join(tmpWorkspace, name), contents);
+            });
+          },
+          tmpWorkspace: async ({}, use) => {
+            // We want a location that won't have a node_modules dir anywhere along its path
+            const tmpWorkspace = await fs.promises.mkdtemp(path.join('/tmp/pwt/workspaces', 'playwright-installation-tests-workspace-'));
+            debug(`Workspace Folder: ${tmpWorkspace}`);
+            await spawnAsync('npm', ['init', '-y'], {
+              cwd: tmpWorkspace,
+            });
+
+            await use(tmpWorkspace);
+            // NOTE WELL: We do not remove for easier debugging
+            // await promisify(rimraf)(tmpWorkspace);
+          },
+          registry: async ({}, use, testInfo) => {
+            const port = testInfo.workerIndex + 16123;
```

Comment:
Not lazy! My old code did that, so it's very possible. However, this is simpler (and we do it for `server` fixture in other projects). If it's dynamic, we have to make `registry.url()` a promise, and rewrite the NPM Metadata's tarball field response at some point since that's generated _before_ we start the server. (We could change it so the server is started before we publish our packages, but I'd rather the server to be down until it's actually ready.)